File: thermal_memo/escpos.py
"""ESC/POS コマンド生成とラスタ画像のパック処理。

このモジュールは GUI にもネットワークにも依存しない純粋なバイト列生成器。
スマホ版（Flutter / Kotlin など）へ移植する際の参照実装でもある。
詳細は docs/PROTOCOL.md を参照。
"""

from __future__ import annotations

from typing import Iterator

ESC = b"\x1b"
GS = b"\x1d"

INIT = ESC + b"@"                     # プリンタ初期化
LEFT_MARGIN_0 = GS + b"L\x00\x00"     # 左マージン 0
ALIGN_LEFT = ESC + b"a\x00"
ALIGN_CENTER = ESC + b"a\x01"
CUT_PARTIAL = GS + b"V\x42\x00"       # Function B: 用紙送り後パーシャルカット


def feed(lines: int) -> bytes:
    """n 行フィード（ESC d n）。"""
    n = max(0, min(255, int(lines)))
    return ESC + b"d" + bytes([n]) if n else b""


def pack_1bit(image) -> tuple[bytes, int, int]:
    """PIL Image を ESC/POS ラスタ用のビット列に変換する。

    戻り値: (data, bytes_per_row, height)
    ビットが 1 の箇所が「黒（印字）」。行末の余りビットは 0（白）で埋める。
    """
    img = image.convert("1")
    width, height = img.size
    bytes_per_row = (width + 7) // 8
    raw = img.tobytes()  # mode "1": 1=白, 行は byte 境界までパディング済み

    # 余りビットを白のまま残すためのマスク（反転後に 0 にする）
    spare = bytes_per_row * 8 - width
    tail_mask = 0xFF if spare == 0 else (0xFF << spare) & 0xFF

    out = bytearray(len(raw))
    for row in range(height):
        start = row * bytes_per_row
        end = start + bytes_per_row
        for i in range(start, end):
            out[i] = (~raw[i]) & 0xFF   # 反転して 1=黒 にする
        out[end - 1] &= tail_mask       # 行末パディングを白へ
    return bytes(out), bytes_per_row, height


def raster_chunks(image, chunk_rows: int = 128) -> Iterator[bytes]:
    """GS v 0 コマンド列を chunk_rows 行ずつ生成する。

    一度に大量の行を送るとプリンタ側の受信バッファを溢れさせる機種があるため、
    帯（バンド）に分割して送る。
    """
    data, bytes_per_row, height = pack_1bit(image)
    chunk_rows = max(1, int(chunk_rows))
    for top in range(0, height, chunk_rows):
        rows = min(chunk_rows, height - top)
        body = data[top * bytes_per_row : (top + rows) * bytes_per_row]
        header = (
            GS + b"v0" + b"\x00"
            + bytes([bytes_per_row & 0xFF, (bytes_per_row >> 8) & 0xFF])
            + bytes([rows & 0xFF, (rows >> 8) & 0xFF])
        )
        yield header + body


def build_job(
    image,
    *,
    copies: int = 1,
    cut: bool = True,
    feed_lines: int = 3,
    chunk_rows: int = 128,
    center: bool = False,
) -> bytes:
    """1 枚分の印刷ジョブ（バイト列）を組み立てる。"""
    parts = [INIT, LEFT_MARGIN_0, ALIGN_CENTER if center else ALIGN_LEFT]
    chunks = list(raster_chunks(image, chunk_rows))
    for index in range(max(1, int(copies))):
        if index:
            parts.append(feed(1))
        parts.extend(chunks)
        parts.append(feed(feed_lines))
        if cut:
            parts.append(CUT_PARTIAL)
    return b"".join(parts)

File: mobile/app/src/main/java/jp/thermalmemo/EscPos.java
package jp.thermalmemo;

import android.graphics.Bitmap;

import java.io.ByteArrayOutputStream;
import java.io.IOException;

/**
 * ESC/POS コマンド生成。
 *
 * <p>デスクトップ版 thermal_memo/escpos.py の移植。仕様は docs/PROTOCOL.md 参照。
 * 印刷対象はすべて 1bit ラスタ画像として送るため、プリンタ内蔵フォントには依存しない。
 */
public final class EscPos {

    public static final byte[] INIT = {0x1B, '@'};
    public static final byte[] LEFT_MARGIN_0 = {0x1D, 'L', 0x00, 0x00};
    public static final byte[] ALIGN_LEFT = {0x1B, 'a', 0x00};
    public static final byte[] ALIGN_CENTER = {0x1B, 'a', 0x01};
    public static final byte[] CUT_PARTIAL = {0x1D, 'V', 0x42, 0x00};

    private EscPos() {
    }

    /** n 行フィード（ESC d n）。 */
    public static byte[] feed(int lines) {
        int n = Math.max(0, Math.min(255, lines));
        if (n == 0) {
            return new byte[0];
        }
        return new byte[]{0x1B, 'd', (byte) n};
    }

    /** 1 行ぶんの ARGB 画素を out の offset 以降へ詰める。 */
    static void packRow(int[] row, int width, byte[] out, int offset) {
        for (int x = 0; x < width; x++) {
            int color = row[x];
            int luminance = (((color >> 16) & 0xFF) * 77
                    + ((color >> 8) & 0xFF) * 150
                    + (color & 0xFF) * 29) >> 8;
            if (luminance < 128) {
                out[offset + (x >> 3)] |= (byte) (0x80 >> (x & 7));
            }
        }
    }

    /**
     * ARGB 配列を「ビット 1 = 黒」のラスタデータへ詰める（Android 非依存。テストもこちらを叩く）。
     *
     * <p>黒い画素にだけビットを立てるので、行末のパディングビットは 0（白）のまま残る。
     * 反転方式で実装すると右端に黒帯が出るので注意（docs/PROTOCOL.md）。
     */
    public static byte[] pack1bit(int[] argb, int width, int height) {
        int bytesPerRow = (width + 7) / 8;
        byte[] out = new byte[bytesPerRow * height];
        int[] row = new int[width];
        for (int y = 0; y < height; y++) {
            System.arraycopy(argb, y * width, row, 0, width);
            packRow(row, width, out, y * bytesPerRow);
        }
        return out;
    }

    /** Bitmap 版。巨大な画像でも一度に全画素を持たないよう 1 行ずつ処理する。 */
    public static byte[] pack1bit(Bitmap bitmap) {
        int width = bitmap.getWidth();
        int height = bitmap.getHeight();
        int bytesPerRow = (width + 7) / 8;
        byte[] out = new byte[bytesPerRow * height];
        int[] row = new int[width];
        for (int y = 0; y < height; y++) {
            bitmap.getPixels(row, 0, width, 0, y, width, 1);
            packRow(row, width, out, y * bytesPerRow);
        }
        return out;
    }

    /**
     * 1 枚分の印刷ジョブを組み立てる。
     *
     * @param chunkRows GS v 0 を分割送信する行数。バッファの小さい機種では小さくする。
     */
    public static byte[] buildJob(Bitmap bitmap, int copies, boolean cut, int feedLines,
                                  int chunkRows, boolean center) throws IOException {
        return assemble(pack1bit(bitmap), bitmap.getWidth(), bitmap.getHeight(),
                copies, cut, feedLines, chunkRows, center);
    }

    /** ARGB 配列から組み立てる（Android 非依存。テストもこちらを叩く）。 */
    public static byte[] buildJob(int[] argb, int width, int height, int copies, boolean cut,
                                  int feedLines, int chunkRows, boolean center) throws IOException {
        return assemble(pack1bit(argb, width, height), width, height,
                copies, cut, feedLines, chunkRows, center);
    }

    private static byte[] assemble(byte[] data, int width, int height, int copies, boolean cut,
                                   int feedLines, int chunkRows, boolean center) throws IOException {
        int bytesPerRow = (width + 7) / 8;

        ByteArrayOutputStream raster = new ByteArrayOutputStream();
        int rowsPerChunk = Math.max(1, chunkRows);
        for (int top = 0; top < height; top += rowsPerChunk) {
            int rows = Math.min(rowsPerChunk, height - top);
            raster.write(new byte[]{
                    0x1D, 'v', '0', 0x00,
                    (byte) (bytesPerRow & 0xFF), (byte) ((bytesPerRow >> 8) & 0xFF),
                    (byte) (rows & 0xFF), (byte) ((rows >> 8) & 0xFF)});
            raster.write(data, top * bytesPerRow, rows * bytesPerRow);
        }
        byte[] band = raster.toByteArray();

        ByteArrayOutputStream job = new ByteArrayOutputStream();
        job.write(INIT);
        job.write(LEFT_MARGIN_0);
        job.write(center ? ALIGN_CENTER : ALIGN_LEFT);
        int total = Math.max(1, copies);
        for (int i = 0; i < total; i++) {
            if (i > 0) {
                job.write(feed(1));
            }
            job.write(band);
            job.write(feed(feedLines));
            if (cut) {
                job.write(CUT_PARTIAL);
            }
        }
        return job.toByteArray();
    }
}
